### PR TITLE
mat: make LU satisfy Matrix

### PR DESCRIPTION
**Pull request by vladimir-ch**
Updates #926

Please take a look.

### CODE REVIEW

File: mat/lu.go [hunk 1]
```diff
@@ -52,37 +106,51 @@ func (lu *LU) updateCond(anorm float64, norm lapack.MatrixNorm) {
...
+// result in the receiver. The LU decomposition will complete regardless of the
+// singularity of a.
 //
-// The LU factorization is computed with pivoting, and so really the decomposition
-// is a PLU decomposition where P is a permutation matrix. The individual matrix
-// factors can be extracted from the factorization using the Permutation method
-// on Dense, and the LU.LTo and LU.UTo methods.
+// The L and U matrix factors can be extracted from the factorization using the
+// LTo and UTo methods. The matrix P can be extracted as a row permutation using
+// the Pivot method and applied using Dense.PermuteRows.
```
Comment (kortschak) [inline mat/lu.go:115]:
```suggestion
// the RowPivots method and applied using Dense.PermuteRows.
```
?

Comment (vladimir-ch) [reply mat/lu.go]: Thanks.

**Comment by vladimir-ch** [comment]
This is now ready.

File: mat/lu.go [hunk 2]
```diff
@@ -19,11 +19,62 @@ const (
...
+//
+// Note that this matrix representation is useful for certain operations, in
+// particular for solving linear systems of equations. It is very inefficient at
+// other operations, in particular At is slow.
 type LU struct {
 	lu    *Dense
 	swaps []int
+	piv   []int
 	cond  float64
+	ok    bool // Whether A is singular
```
Comment (kortschak) [inline mat/lu.go:40]: This is a little confusing; the intention is that `!ok` → singular, but it could be read differently.

**Review by kortschak — COMMENTED** [review]
Minor nit only.

File: mat/lu.go (hunk 2 above)
Comment (vladimir-ch) [reply mat/lu.go]: Good point, thanks. I've also adjusted comments in the involved Lapack routines.

**Review by kortschak — APPROVED** [review]

**Pull request merged**